go/libraries/doltcore/doltdb/root_val.go

```diff
@@ -484,50 +492,154 @@ func (root *RootValue) RemoveTables(ctx context.Context, tables ...string) (*Roo
 	return newRootValue(root.vrw, rootValSt), nil
 }
 
-// DocDiff returns the slices of the dolt docs on the filesystem that are added, modified, and removed when compared with a root value.  Docs
-// In this instance that are not in the other instance are considered added, and docs in the other instance and not
-// this instance are considered removed.
-func (root *RootValue) DocDiff(ctx context.Context, docDetails []*DocDetails) (added, modified, removed []string, err error) {
-	docTable, found, err := root.GetTable(ctx, docTableName)
+// DocDiff returns the added, modified and removed docs when comparing a root value with an other (newer) value. If the other value,
+// is not provided, then we compare the docs on the root value to the docDetails provided.
+func (root *RootValue) DocDiff(ctx context.Context, other *RootValue, docDetails []*DocDetails) (added, modified, removed []string, err error) {
+	added = []string{}
+	modified = []string{}
+	removed = []string{}
 
+	oldTbl, oldTblFound, err := root.GetTable(ctx, DocTableName)
 	if err != nil {
 		return nil, nil, nil, err
 	}
+	var oldSch schema.Schema
+	if oldTblFound {
+		sch, err := oldTbl.GetSchema(ctx)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		oldSch = sch
+	}
+
+	if other == nil {
+		if oldTblFound {
+			for i, details := range docDetails {
+				pkTaggedVal := row.TaggedValues{
+					DocNameTag: types.String(details.DocPk),
+				}
+
+				docRow, ok, err := oldTbl.GetRowByPKVals(ctx, pkTaggedVal, oldSch)
+				if err != nil {
+					return nil, nil, nil, err
+				}
+
+				if ok {
+					docValue, _ := docRow.GetColVal(DocTextTag)
+					docDetails[i] = &DocDetails{details.NewerText, details.DocPk, docValue}
+				}
+			}
+		}
+		for _, doc := range docDetails {
+			added, modified, removed = appendDocDiffs(added, modified, removed, doc.Value, doc.NewerText, doc.DocPk)
+		}
+		return added, modified, removed, err
+	}
+
+	newTbl, newTblFound, err := other.GetTable(ctx, DocTableName)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	var newSch schema.Schema
+	if newTblFound {
+		sch, err := newTbl.GetSchema(ctx)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		newSch = sch
+	}
 
-	var rowMap types.Map
-	if found {
-		rowMap, err = docTable.GetRowData(ctx)
+	if newTblFound {
+		newRows, err := newTbl.GetRowData(ctx)
 		if err != nil {
 			return nil, nil, nil, err
 		}
+		err = newRows.IterAll(ctx, func(key, val types.Value) error {
```

Comment:
@bheni Lmk if you'd recommend a better way of doing this.... right now I'm iterating through all the newRows (to find modified and appended) and then all the oldRows (to find deleted). I thought about nesting the iterators but didn't want to increase the complexity.

Reply:
I think the approach is fine.  But this function has gotten a bit large, and hard to understand what is going on.  Would be clearer if you broke it up into functions with names that clearly describe each stage.

Reply:
I broke up the root.DocDiff function quite a bit. I could still probably extract one more helper function from within these iterators. I attempted it today, but got stuck and needed to move on.